File: BddPipe/BddPipe.UnitTests/StringExtensionsTests.cs
using BddPipe.UnitTests.Asserts;
using FluentAssertions;
using NUnit.Framework;
using static BddPipe.F;

namespace BddPipe.UnitTests
{
    [TestFixture]
    public class StringExtensionsTests
    {
        private const string DefaultValue = "test";
        private const string DefaultPrefix = "Given";
        private const string DefaultExpectation = "Given test";

        [Test]
        public void WithPrefix_DoesNotHavePrefix_PrefixAdded()
        {
            var result = DefaultValue.AsOption().WithPrefix(DefaultPrefix);
            result.Value.Should().Be(DefaultExpectation);
        }

        [Test]
        public void WithPrefix_AlreadyHasPrefix_PrefixNotAdded()
        {
            var result = DefaultExpectation.AsOption().WithPrefix(DefaultPrefix);
            result.Value.Should().Be(DefaultExpectation);
        }

        [Test]
        public void WithPrefix_AlreadyHasPrefixCaseLower_PrefixNotAdded()
        {
            var result = "given something".AsOption().WithPrefix(DefaultPrefix);
            result.Value.Should().Be("given something");
        }

        [Test]
        public void WithPrefix_AlreadyHasPrefixCaseUpper_PrefixNotAdded()
        {
            var result = "GIVEN something".AsOption().WithPrefix(DefaultPrefix);
            result.Value.Should().Be("GIVEN something");
        }

        [Test]
        public void WithPrefix_AlreadyHasPrefixCaseMixed_PrefixNotAdded()
        {
            var result = "GiVeN something".AsOption().WithPrefix(DefaultPrefix);
            result.Value.Should().Be("GiVeN something");
        }

        [Test]
        public void WithPrefix_Null_PrefixReturned()
        {
            var result = ((string)null).AsOption().WithPrefix(DefaultPrefix);
            result.Value.Should().Be(DefaultPrefix);
        }

        [Test]
        public void WithPrefix_Empty_PrefixReturned()
        {
            var result = string.Empty.AsOption().WithPrefix(DefaultPrefix);
            result.Value.Should().Be(DefaultPrefix);
        }

        [Test]
        public void WithPrefix_Whitespace_PrefixReturned()
        {
            var result = " ".AsOption().WithPrefix(DefaultPrefix);
            result.Value.Should().Be(DefaultPrefix);
        }

        [Test]
        public void NoneIfWhiteSpace_OptionIsNone_None()
        {
            Option<string> source = None;

            source.NoneIfWhiteSpace().ShouldBeNone();
        }

        [Test]
        public void NoneIfWhiteSpace_OptionIsSomeText_SomeText()
        {
            Option<string> source = "value";

            source.NoneIfWhiteSpace().ShouldBeSome(txt =>
            {
                txt.Should().Be("value");
            });
        }

        [Test]
        public void NoneIfWhiteSpace_OptionIsSomeTextTrailingWhiteSpace_SomeTextDoesNotTrim()
        {
            Option<string> source = "value ";

            source.NoneIfWhiteSpace().ShouldBeSome(txt =>
            {
                txt.Should().Be("value ");
            });
        }

        [Test]
        public void NoneIfWhiteSpace_OptionIsEmpty_None()
        {
            Option<string> source = string.Empty;

            source.NoneIfWhiteSpace().ShouldBeNone();
        }

        [TestCase(" ")]
        [TestCase("  ")]
        [TestCase("\t")]
        [TestCase("\t ")]
        [TestCase(" \t")]
        [TestCase("\r\n")]
        [TestCase(" \t\r\n")]
        public void NoneIfWhiteSpace_OptionIsWhiteSpace_None(string value)
        {
            Option<string> source = value;

            source.NoneIfWhiteSpace().ShouldBeNone();
        }
    }
}

File: BddPipe/BddPipe.FSharp.UnitTests/StringExtensionsTests.fs
namespace BddPipe.FSharp.UnitTests

open NUnit.Framework
open BddPipe.FSharp.StringExtensions
open FsUnit

[<TestFixture>]
module StringExtensionsTests =

    [<Literal>]
    let DefaultValue = "test";

    [<Literal>]
    let DefaultPrefix = "Given";

    [<Literal>]
    let DefaultExpectation = "Given test";

    [<Test>]
    let WithPrefix_DoesNotHavePrefix_PrefixAdded() =
        let result = Some DefaultValue |> withPrefix DefaultPrefix
        result |> should equal DefaultExpectation

    [<Test>]
    let WithPrefix_AlreadyHasPrefix_PrefixNotAdded() =
        let result = Some DefaultExpectation |> withPrefix DefaultPrefix
        result |> should equal DefaultExpectation

    [<Test>]
    let WithPrefix_AlreadyHasPrefixCaseLower_PrefixNotAdded() =
        let result = Some "given something" |> withPrefix DefaultPrefix
        result |> should equal "given something"

    [<Test>]
    let WithPrefix_AlreadyHasPrefixCaseUpper_PrefixNotAdded() =
        let result = Some "GIVEN something" |> withPrefix DefaultPrefix
        result |> should equal "GIVEN something"

    [<Test>]
    let WithPrefix_AlreadyHasPrefixCaseMixed_PrefixNotAdded() =
        let result = Some "GiVeN something" |> withPrefix DefaultPrefix
        result |> should equal "GiVeN something"

    [<Test>]
    let WithPrefix_None_PrefixReturned() =
        let result = None |> withPrefix DefaultPrefix
        result |> should equal DefaultPrefix

    [<Test>]
    let WithPrefix_Empty_PrefixReturned() =
        let result = Some "" |> withPrefix DefaultPrefix
        result |> should equal DefaultPrefix

    [<Test>]
    let WithPrefix_Whitespace_PrefixReturned() =
        let result = Some " " |> withPrefix DefaultPrefix
        result |> should equal DefaultPrefix
